[Gradient Compression] Add get_per_parameter_tensors method to GradBucket class

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53102 [Gradient Compression] Don't provide default values in GradBucket constructor
* **#53009 [Gradient Compression] Add get_per_parameter_tensors method to GradBucket class**
* #53008 Clang-format c10d/init.cpp

It can be a common operation to apply layer-wise operations over per-parameter tensors in a DDP communication hook.

Create a util method in GradBucket class before publishing GradBucket APIs.

Differential Revision: [D26717893](https://our.internmc.facebook.com/intern/diff/D26717893/)

Commit:
[Gradient Compression] Add get_per_parameter_tensors method to GradBucket class

It can be a common operation to apply layer-wise operations over per-parameter tensors in a DDP communication hook.

Create a util method in GradBucket class before publishing GradBucket APIs.

Differential Revision: [D26717893](https://our.internmc.facebook.com/intern/diff/D26717893/)

[ghstack-poisoned]

rohan-varma commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -18,10 +18,11 @@ def _register_builtin_comm_hook(reducer: Reducer, comm_hook_type: BuiltinCommHoo
+    def get_tensors(self) -> List[Tensor]: ...
+    def get_per_parameter_tensors(self) -> List[Tensor]: ...
     def get_offsets(self) -> List[int]: ...
+    def get_lengths(self) -> List[int]: ...
```
Are `get_offsets`, `get_sizes`, and `get_lengths` no longer intended to be used by GradBucket class user? If so, consider making those private APIs?

SciPioneer replied:
You are right.

Actually I removed them in a follow-up PR. I don't think we will use them.

rohan-varma commented on torch/lib/c10d/comm.cpp:
```diff
@@ -87,4 +86,17 @@ void broadcast_coalesced(
 
+std::vector<at::Tensor> GradBucket::getPerParameterTensors() const {
+  std::vector<at::Tensor> per_parameter_tensors;
+  size_t num_parameters = offsets_.size();
```
It looks like in constructor, offsets, lengths, and sizes_vec are optional arguments and could possibly be empty. It looks like only the reducer initializes the GradBucket and passes in these values, but I'm wondering if we should throw an error here if those arguments were not specified. For example I think that this would return just an empty vector if `offsets_` wasn't provided, which is probably not the result we want to return.

SciPioneer replied:
Good point! Let me address this in a follow-up PR, as it's a separate topic. I plan to remove these default values and polish the constructor API.

Previously I set some default values probably because it can minimize the changes in some references (probably some tests).

rohan-varma commented on torch/lib/c10d/comm.hpp:
```diff
@@ -49,6 +49,9 @@ class GradBucket {
   }
 
+  // Each tensor in the list that getPerParameterTensors corresponds to a parameter.
+  std::vector<at::Tensor> getPerParameterTensors() const;
```
Could we add python unittests that ensure `getPerParameter()` succeeds, i.e. compare it to the tensor for a given parameter in python model?

SciPioneer replied:
I don't think it's necessary since the caller powerSGD_hook is involved in the unit test, and I have also run PowerSGD e2e to make sure no accuracy/performance regression.

Commit:
Update on "[Gradient Compression] Add get_per_parameter_tensors method to GradBucket class"


It can be a common operation to apply layer-wise operations over per-parameter tensors in a DDP communication hook.

Create a util method in GradBucket class before publishing GradBucket APIs.

Differential Revision: [D26717893](https://our.internmc.facebook.com/intern/diff/D26717893/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -199,18 +199,33 @@ PyObject* c10d_init(PyObject* _unused, PyObject* noargs) {
+    Since the buckets are rebuilt after the first iteration, should not rely on the indices at the beginning of training.
+
+Returns:
+    The index of a bucket that stores gradients of a few contiguous layers.
```
This can be addressed in a separate PR, but I'm not sure if we should keep `get_Index()` as a public method since gradient bucketing/rebuilding buckets is sort of a DDP internal detail that shouldn't leak to the user.

I see that this is mainly used as key into an error_dict, can we provide a way to hash `GradBucket` instead to use that as a key to the error_dict?

SciPioneer replied:
Agree that it may unnecessarily expose some internal details.

`bucket_index == 0` is also used to determine if the current bucket is the last one to allreduce in an iteration. This is something hard to be done by a hashcode.

rohan-varma reviewed (approved):
LGTM, although please consider the inline comment. Also, are we going to make the APIs such as `get_offsets`, `get_lengths` etc private?
